src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -30,10 +30,25 @@ public partial class Solution
         // Values for all these are created on demand.
         private ImmutableHashMap<ProjectId, Project> _projectIdToProjectMap;
 
+        /// <summary>
+        /// Result of calling <see cref="WithFrozenPartialCompilationsAsync"/>.
+        /// </summary>
+        private AsyncLazy<Solution> _cachedFrozenSolution;
+
+        /// <summary>
+        /// Mapping of DocumentId to the frozen solution we produced for it the last time we were queried.  Use <see
+        /// cref="_cachedFrozenGate"/> to access.
+        /// </summary>
+        private readonly Dictionary<DocumentId, Solution> _cachedFrozenDocumentState = new();
```

Comment:
>_cachedFrozenDocumentState [](http://example.com/codeflow?start=58&length=26)

I don't think I'm clear on why both Solution and SolutionCompilationState need this Dictionary. I'm sure there's a good reason, I just don't understand this well enough yet to reason out why.

Reply:
i was being paranoid, just to make sure that no matter how we may have gotten the Solution or SolutionCompilationState that we had the caching.  We likely only need the outer one.